hedera-node/hedera-util-service-impl/src/main/java/com/hedera/node/app/service/util/impl/handlers/UtilPrngHandler.java

```diff
@@ -18,27 +18,144 @@
+        // from the pseudoRandomBytes
+        if (range > 0) {
+            final int pseudoRandomNumber = randomNumFromBytes(pseudoRandomBytes, range);
+            recordBuilder.setPrngNumber(pseudoRandomNumber);
+        } else {
+            recordBuilder.setPrngBytes(Bytes.wrap(pseudoRandomBytes));
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public PrngRecordBuilder newRecordBuilder() {
+        return new UtilPrngRecordBuilder();
+    }
+
+    /**
+     * Get the n-3 running hash bytes from the n-3 running hash. If the running hash is not
+     * available, will throw a HandleException. n-3 running hash is chosen for generating random
+     * number instead of n-1 running hash for processing transactions quickly. Because n-1 running
+     * hash might not be available when the transaction is processed and might need to wait longer.
+     *
+     * @param nMinus3RunningHash n-3 running hash
+     * @return n-3 running hash bytes
+     */
+    private byte[] getNMinus3RunningHashBytes(@NonNull final RunningHash nMinus3RunningHash) {
+        // This can't happen because this running hash is taken from record stream.
+        // If this happens then there is a bug in the code.
+        requireNonNull(nMinus3RunningHash);
+
+        Hash nMinusThreeHash;
+        try {
+            nMinusThreeHash = nMinus3RunningHash.getFutureHash().get();
+            // Use n-3 running hash instead of n-1 running hash for processing transactions quickly
+            if (nMinusThreeHash == null || Arrays.equals(nMinusThreeHash.getValue(), new byte[48])) {
+                log.info("No n-3 record running hash available to generate random number");
+                return MISSING_BYTES;
+            }
+            // generate binary string from the running hash of records
+            return nMinusThreeHash.getValue();
+        } catch (InterruptedException e) {
+            log.error("Interrupted exception while waiting for n-3 running hash", e);
+            Thread.currentThread().interrupt();
+            throw new HandleException(ResponseCodeEnum.UNKNOWN);
+            // FUTURE : Need to decide on the response code for this case
+        } catch (ExecutionException e) {
+            log.error("Unable to get current n-3 running hash", e);
+            throw new HandleException(ResponseCodeEnum.UNKNOWN);
+            // FUTURE : Need to decide on the response code for this case
+        }
+    }
+
+    /**
+     * Generate a random number from the given bytes in the given range
+     * @param pseudoRandomBytes bytes to generate random number from
+     * @param range range of the random number
+     * @return random number
+     */
+    private int randomNumFromBytes(@NonNull final byte[] pseudoRandomBytes, final int range) {
+        final int initialBitsValue = ByteBuffer.wrap(pseudoRandomBytes, 0, 4).getInt();
+        return IntMath.mod(initialBitsValue, range);
```

Comment:
Is there a reason we used Google's IntMath class instead of the modulus operator (`%`)?

Reply:
If I remember correctly,  it was suggested by Leemon when the HIP was implemented.

Reply:
It's extra dependency and slower than the operator, but probably doesn't matter in this method.  The biggest difference is that Google's method would throw ArithmeticException if range was negative and flip the result if initialBitsValue was negative.  Internally it still calls the operator.